### PR TITLE
Allow to import state roots from dump

### LINKED ISSUE #1307: Stateroot export/import support
There is a format now for stateroot data dumps, see neo-project/neo-modules#315. We should add support for that.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1307 .

Testing done:
1. Imported test dump, and exported it with roots.
2. Imported test dump with roots with no errors, including incremental import (`--count 1 --start x`).
3. Block dump files exported with and without `--state` argument don't differ.

### CODE REVIEW

**Comment by roman-khimov** [comment]
I think we can wait for some official dumps to test this.

File: cli/server/server.go [hunk 1]
```diff
@@ -65,6 +66,10 @@ func NewCommands() []cli.Command {
 			Name:  "diff, k",
 			Usage: "Use if DB is restore from diff and not full dump",
 		},
+		cli.BoolFlag{
+			Name:  "state, r",
+			Usage: "Import state roots instead of blocks",
```
Comment (roman-khimov) [inline cli/server/server.go:75]: Can we import both at the same time? Seems like a valid use case.

File: cli/server/server.go [hunk 2]
```diff
@@ -50,6 +50,10 @@ func NewCommands() []cli.Command {
 			Name:  "out, o",
 			Usage: "Output file (stdout if not given)",
 		},
+		cli.BoolFlag{
+			Name:  "state, r",
+			Usage: "Export state roots instead of blocks",
```
Comment (roman-khimov) [inline cli/server/server.go:55]: Both?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: cli/server/server.go (hunk 1 above)
Comment (fyrchik) [reply cli/server/server.go]: Fixed.

File: cli/server/server.go (hunk 2 above)
Comment (fyrchik) [reply cli/server/server.go]: Fixed.

**Comment by fyrchik** [comment]
Now we can optionally export/import roots together with blocks.

File: cli/server/server.go [hunk 3]
```diff
@@ -285,13 +300,38 @@ func restoreDB(ctx *cli.Context) error {
...
+		rootStart = rootsIn.ReadU32LE()
+		rootSize = rootsIn.ReadU32LE()
+		if rootsIn.Err != nil {
+			return cli.NewExitError(fmt.Errorf("error while reading roots file: %w", rootsIn.Err), 1)
+		}
+		if start < rootStart {
+			return cli.NewExitError(fmt.Errorf("roots file start from %d root, can't import %d", rootStart, start), 1)
+		}
+		lastRoot := rootStart + rootSize
+		if rootStart+count > lastRoot {
```
Comment (roman-khimov) [inline cli/server/server.go:348]: I'm not sure it's correct treating it this way. State root chain is almost always one block behind and for a dump of blocks up to N height it's normal to only have dump of states up to N-1 height. This should be accounted for.

File: cli/server/server.go [hunk 4]
```diff
@@ -209,22 +213,42 @@ func dumpDB(ctx *cli.Context) error {
...
+		rootsWriter = io.NewBinWriterFromIO(rootsStream)
+	}
 	if start != 0 {
 		writer.WriteU32LE(start)
 	}
 	writer.WriteU32LE(count)
+
+	if rootsWriter != nil {
+		rootsWriter.WriteU32LE(start)
+		rootsWriter.WriteU32LE(count)
```
Comment (roman-khimov) [inline cli/server/server.go:233]: Same thing here, do we really have this `count` of states when exporting whole chain we have?

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**